mcase/lib.sh

```diff
@@ -0,0 +1,1416 @@
+#
+#   You should have received a copy of the GNU General Public License
+#   along with this program. If not, see http://www.gnu.org/licenses/.
+#
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#   library-prefix = distribution_mcase__
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+#
+# Modal test case
+#
+# This module brings modal approach to defining and running test
+# code.
+#
+# Instead of writing your test case as a single body of code,
+# you will define handlers for different modes: setup, test, diag
+# and cleanup, and call a special function which will run all the
+# handlers in particular order.
+#
+# The standard order is `setup`, `test`, `diag`, `cleanup` but the
+# runner function can amend this order based on instructions from test
+# scheduler, typically passed as environment variables.
+#
+# Typical example of this is an upgrade test that will run `setup` and
+# `test` on first environment, then upgrade the system and run `test`
+# on upgraded system again.
+#
+#
+# =head1 HELP, MY TEST IS FAILING AND I DON'T UNDERSTAND THIS SORCERY!
+#
+# If you encounter mcase-based test and want to understand what's
+# happening, usually the fastest way to get the grip is:
+#
+#  1. Look at early log text around 'selected workflow:'  This is
+#     where the "workflow" is shown, ie. planned sequence of handlers
+#
+#  2. All the test really does is just call these handlers in that order,
+#     wrapping them in phases.
+#
+#  4. Open the test code.
+#
+#     For every phase (setup, test, diag and cleanup), there is
+#     one function ("handler") named `distribution_mcase__test`
+#     etc. (Actually only 'test' is mandatory.)
+#
+#
+# =head1 GETTING STARTED
+#
+# Here's what you need to do:
+#
+#  1. Implement handlers:
+#
+#         distribution_mcase__setup (optional)
+#         distribution_mcase__test
+#         distribution_mcase__diag (optional)
+#         distribution_mcase__cleanup (optional)
+#
+#  2. Finally, run a single "magic" function, `distribution_mcase__run()`.
+#     This will take care of the rest.
+#
+#
+# =head1 EXAMPLE
+#
+#     distribution_mcase__setup() {
+#         rlRun "mkdir /var/ftp"                    || return 1
+#         rlRun "cp testfile /var/ftp"              || return 1
+#         rlRun "useradd joe"                       || return 1
+#         rlRun "rlServiceRestart hypothetical_ftp  || return 1
+#         rlRun "rlServiceEnable hypothetical_ftp   || return 1
+#     }
+#
+#     distribution_mcase__test() {
+#         rlRun "su -c 'hypo_ftp localhost 25 <<<\"get testfile\"' - joe"
+#         rlRun "diff /var/ftp/testfile /home/joe/testfile"
+#     }
+#
+#     distribution_mcase__cleanup() {
+#         rlRun "rm -rf /var/ftp"
+#         rlRun "userdel joe"
+#         rlRun "rlServiceRestart hypothetical_ftp
+#     }
+#
+#
+#     rlJournalStart
+#
+#         rlPhaseStartSetup
+#             rlImport distribution/mcase
```

Comment:
```suggestion
#             rlImport ControlFlow/mcase
```
this would be preferred, however after implementing the _distribution/mcase_ forwarder it should be fine as well